libs/s25main/EconomyModeHandler.cpp

```diff
@@ -0,0 +1,302 @@
+
+        gwg->GetPostMgr().SetMissionGoal(p, goalText);
+    }
+}
+
+EconomyModeHandler::EconomyModeHandler(SerializedGameData& sgd, unsigned objId)
+    : GameObject(sgd, objId), endFrame(sgd.PopUnsignedInt()), gfLastUpdated(0)
+{
+    sgd.PopContainer(goodsToCollect);
+
+    std::vector<unsigned> teamBitMasks;
+    sgd.PopContainer(teamBitMasks);
+    for(auto& teamMask : teamBitMasks)
+    {
+        economyModeTeams.emplace_back(teamMask, goodsToCollect.size());
+    }
+
+    amountsThePlayersCollected.resize(goodsToCollect.size());
+    maxAmountsATeamCollected.resize(goodsToCollect.size());
+}
+
+void EconomyModeHandler::Destroy() {}
+
+/// Serialisierungsfunktion
+void EconomyModeHandler::Serialize(SerializedGameData& sgd) const
+{
+    sgd.PushUnsignedInt(endFrame);
+    sgd.PushContainer(goodsToCollect);
+    std::vector<unsigned> teamBitMasks;
+    for(const EconomyModeHandler::EconTeam& curTeam : economyModeTeams)
+    {
+        teamBitMasks.push_back(curTeam.playersInTeam.to_ulong());
+    }
+    sgd.PushContainer(teamBitMasks);
+}
+
+void EconomyModeHandler::DetermineTeams()
+{
+    for(unsigned i = 0; i < gwg->GetNumPlayers(); ++i)
+    {
+        if(gwg->GetPlayer(i).isUsed())
+        {
+            bool foundTeam = false;
+            for(const auto& team : economyModeTeams)
+            {
+                if(team.containsPlayer(i))
+                {
+                    foundTeam = true;
+                    break;
+                }
+            }
+            if(!foundTeam)
+            {
+                const GamePlayer& player = gwg->GetPlayer(i);
+                std::bitset<MAX_PLAYERS> newTeam;
+                newTeam.set(i);
+                for(unsigned j = i + 1; j < gwg->GetNumPlayers(); ++j)
+                {
+                    if(gwg->GetPlayer(j).isUsed() && player.IsAlly(j))
+                    {
+                        newTeam.set(j);
+                    }
+                }
+                economyModeTeams.emplace_back(newTeam, goodsToCollect.size());
+            }
+        }
+    }
+}
+
+unsigned EconomyModeHandler::SumUpGood(GoodType good, const Inventory& Inventory)
+{
+    unsigned retVal = Inventory.goods[good];
+
+    // Add the tools used by workers to the good totals
+    for(unsigned j = 0; j < NUM_JOB_TYPES; j++)
+    {
+        boost::optional<GoodType> tool = JOB_CONSTS[(Job)j].tool;
+        if(tool == good)
+        {
+            retVal += Inventory.people[j];
+        }
+    }
+    // Add the weapons and beer used by soldiers to the good totals
+    if(good == GD_BEER || good == GD_SWORD || good == GD_SHIELDROMANS)
+    {
+        for(const auto& it : SOLDIER_JOBS)
+        {
+            retVal += Inventory.people[it];
+        }
+    }
+
+    return retVal;
+}
+
+void EconomyModeHandler::UpdateAmounts()
+{
+    // Return if the game is over or we already updated the amounts this game frame
+    if(isOver() || gfLastUpdated == GetEvMgr().GetCurrentGF())
```

Comment:
do we want to update each GF? or should we also say its calculated every "now and then"?

Reply:
besides this question everything looks good!

Reply:
If you go for the "now and then" approach: I remember that there is a "statistics step". I guess that would be a good occasion.

@Flow86 Is the "heavy load" measured? If so, maybe we can improve that?
BTW: Do our toolchains support OpenMP? That would be a simple way to get some more speed here.

Reply:
I looked at this. Indeed, if I open the Economic progress window the framerate (without frame limiting) drops significantly. So I analysed and optimized a bit and even changed it to only update the iwEconomicProgress content at most once every second, because that's certainly enough.

But it doesn't make a significant difference in the frame rate. Indeed, when you do some perfomance analysis, you will see that the only thing for the ingame windows that takes any significant amount of time at all is drawing. And this seems to depend on the number and type of control elements displayed in the window, for example the tool window with tool ordering enabled is even much worse than the economic progress window. So I recommend that if the goal is to reduce the load while displaying ingame windows the Draw routine should be optimized, for example by prerendering/baking the whole windows except elements that change into textures or if possible by using vector buffers instead of directly drawing quads. But this is vastly beyond the scope of this pull request.

So tldr: I checked, and no, in comparison with the whole situation of drawing an ingame window these calculations are not a heavy load at all.

Reply:
Should be all good now again

Reply:
I thought about this again, and to be honest, I don't like these last three "performance optimization" commits I made very much, because of stylistic and clarity reasons, and as I wrote it's probably unnecessary to have them. I regret them and think I will revert them tomorrow. Should also make it easier to review as @Flow86 already said before that it basically looks good.